modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/QueueListener.java

```diff
@@ -18,30 +18,75 @@
 
 package org.wso2.andes.kernel;
 
+import org.apache.log4j.Logger;
+import org.wso2.andes.server.ClusterResourceHolder;
+import org.wso2.andes.server.cluster.coordination.ClusterEventNotifier;
+import org.wso2.andes.server.cluster.coordination.ClusterNotification;
+
 /**
- * Listener listening for queue changes local and cluster
+ * QueueListener listens and handles queue changes that have occurred locally and cluster-wide.
  */
-public interface QueueListener {
+public class QueueListener {
+
+    Logger log = Logger.getLogger(QueueListener.class);
 
-    public static enum QueueEvent {
+    public enum QueueEvent {
         ADDED,
         DELETED,
         PURGED
     }
 
     /**
-     * Handle queue has changed in the cluster
+     * Local event handler is used to notify local queue changes to the cluster.
+     */
+    private ClusterEventNotifier clusterEventNotifier;
+
+    public QueueListener(ClusterEventNotifier handler) {
+        clusterEventNotifier = handler;
+    }
+
+    /**
+     * Handle when a queue has changed in the cluster.
      *
      * @param andesQueue changed queue
      * @param changeType what type of change has happened
      */
-    public void handleClusterQueuesChanged(AndesQueue andesQueue, QueueEvent changeType) throws AndesException;
+    public void handleClusterQueuesChanged(AndesQueue andesQueue, QueueEvent changeType) throws AndesException {
+        if (log.isDebugEnabled()) {
+            log.debug("Cluster event received: " + andesQueue.encodeAsString());
+        }
+        switch (changeType) {
+            case ADDED:
+                //create a queue
+                ClusterResourceHolder.getInstance().getVirtualHostConfigSynchronizer().clusterQueueAdded(andesQueue);
+                break;
+            case DELETED:
+                //Delete remaining subscriptions from the local and cluster subscription maps
+                ClusterResourceHolder.getInstance().getSubscriptionManager().deleteAllLocalSubscriptionsOfBoundQueue(
+                        andesQueue.queueName, andesQueue.getProtocolType(), andesQueue.getDestinationType());
```

Comment:
for 'handleClusterQueuesChanged' to run, somebody should have called 'handleLocalQueuesChanged' before that. If this is the same node, the above said problem will occur. If this was generated by another node, the cluster event listener task only starts after initializing the the subscription manager so events won't be receive before initializing the sub manager.